### ISSUE TITLE
Non-deterministic file reference lists are generated

### ISSUE THREAD

**Issue by igor-makarov** [jcampbell05/xcake#172]
I'm researching why two subsequent iterations of xcake produce completely different project files - they look the same as far as Xcode group sort order is concerned, but the PBX file is completely different.

I know that Cocoapods somehow manages to avoid this problem when generating the Pods.xcodeproj.

In xcake, what determines the order of the files traversed? It seems like something is being added in a non-deterministic manner.

**Comment by jcampbell05** [comment]
@igor-makarov we collect them using the `glob` method in Ruby. We traverse them according to the order provided in that.

I haven't touched it for a while but its possible that during adding them the code that figures out how to do that doesn't keep the UUID consistent or something ?

**Comment by igor-makarov** [comment]
I've checked and the glob does return consistent sorting. However, the order in which the files appear in the pbxproj is different for each generation.

**Comment by igor-makarov** [comment]
I've tried using `xcodeproj.sort` and xUnique, but it's not helping because the differences are too big.

**Comment by yonaskolb** [comment]
Are the reference ids different? This is usually the cause for generated xcodeproject diffs

**Comment by igor-makarov** [comment]
UUIDs are different, of course. But this goes deeper, as the actual order of the refs is different, which causes `xcodeproj.predictabilize_uuids` to produce different deterministic UUIDs for each run - it relies on applying an MD5 on the full tree-paths for each node.

**Comment by yonaskolb** [comment]
Ok, that's another issue then.
Not sure if it fits your needs like xcake does but for what it's worth https://github.com/yonaskolb/XcodeGen has deterministic generation

**Comment by jcampbell05** [comment]
@igor-makarov yeah I am not sure, at this point. It goes into a generator system potentially it adds the files to the project in a different order. I think the order in which they added is chronological to when the add_file is called.

**Comment by igor-makarov** [comment]
What affects the order in which the add_file is called?

**Comment by jcampbell05** [comment]
@igor-makarov right now the generators declare their dependencies on each other which determines the order they are called in. Potentially one generator wants to add a file to the project perhaps it handles the `xcconfigs` for example then it will add those files first.

Perhaps the generator used for adding all of the files should always be ran first (but after the project and targets are ready)

**Comment by igor-makarov** [comment]
I checked, the order in which the `include_file_for_path_and_target` are called on files doesn't change. It's something down the line that clobbers the PBX node order.

**Comment by jcampbell05** [comment]
@igor-makarov Maybe check to see that nothing decides to sort them alphabetically or something nieave like that.

**Comment by igor-makarov** [comment]
I wish! That would have been better, actually. It's literally generating a different order run after run, even if no files change.

**Comment by igor-makarov** [comment]
BTW, I 'solved' the problem of getting different output on same input like so:
I added:
```
  project.sort
  project.predictabilize_uuids
```
in the `after_save` phase. Perhaps this should be added into Xcake itself?

Anyway, adding or removing a file still upends the entire PBX tree child order and produces a terrible long diff.

It's not like it's even a bug - I know that when using Xcake and Cocoapods I shouldn't care what's inside the PBXs. But I still want them committed to git, so that switching branches doesn't necessitate running ./init scripts every time. And this means diff towers of doom.

In conclusion, PBX needs to be destroyed.

Thanks for Xcake, @jcampbell05 - I'm definitely better off with it than without it.

**Comment by jcampbell05** [comment]
@igor-makarov interesting We should try that out internally and see if that makes a difference.

I guess if there was a way Xcake could consume an existing PBX and diff it manually so that files and their UUIDs are preserved. Currently it re-writes it everytime, I'm not sure how much work this would be.

Cheers :) thanks for the contributions so far, been meaning to add some of this stuff but didn't get the time.

**Comment by igor-makarov** [comment]
My final take on it is that it's not really necessary :/

The bottom line in my opinion is that we shouldn't care about the PBX file contents as long as it performs the right build steps.

**Comment by jcampbell05** [comment]
Okay so just 
```
 project.sort
 project.predictabilize_uuids
```

**Comment by igor-makarov** [comment]
Yeah I guess. But not even that helps prevent clobbering...

**Comment by jcampbell05** [comment]
So close for now until we have more information ?

**Comment by igor-makarov** [comment]
Or until PBX is dead, whichever comes first, LOL

**Comment by jcampbell05** [comment]
I think the Swift Project Manager will show it the door

**Comment by igor-makarov** [comment]
Too bad SwiftPM won't build my huge pile of ObjC files...

**Comment by jcampbell05** [comment]
@igor-makarov that's the apple way sadly :(

**Comment by igor-makarov** [comment]
I just want to never look at a PBX diff ever again :/
